File: #14_Longest_Common_Prefix/Solution.py
'''
1) "Horizontal matching (over strings)". 
Pick up the first string and compare it with the rest. Return the minimum prefix found among all comparisons.

2) "Vertical matching (over characters)". 
Compare the characters between all strings from left to right. Stop whenever a mismatch is found.

Second one is faster because it is faster to fail.
Worst case time complexity is the same. O(n * m) where n is the length of array and m is average strlen
'''

class Solution:
    # @param {string[]} strs
    # @return {string}
    def longestCommonPrefix(self, strs):
        if not strs:
        	return ''
        lcp = 0
        while lcp < len(strs[0]):
        	for i in range(1, len(strs)):
        		if lcp >= len(strs[i]) or strs[0][lcp] != strs[i][lcp]:
        			return strs[0][:lcp]
        	lcp += 1
        return strs[0]

sol = Solution()
strs = ["abc", "ab", "abcde"]
print sol.longestCommonPrefix(strs)

File: #14_Longest_Common_Prefix/Solution.java
public class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) return "";
        int i = 0; // length of lcp
        for (; i < strs[0].length(); i++)
        	for (int j = 1; j < strs.length; j++)
        		if (i == strs[j].length() || strs[0].charAt(i) != strs[j].charAt(i)) return strs[0].substring(0, i);
        return strs[0].substring(0, i);
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	String[] strs = {"aa", "a"};
    	System.out.println(sol.longestCommonPrefix(strs));
    }
}